core/src/main/java/io/questdb/Server.java

```diff
@@ -0,0 +1,210 @@
+                config.getHttpMinServerConfiguration(),
+                workerPoolManager,
+                cairoEngine,
+                metrics
+        ));
+
+        // pg wire
+        if (config.getPGWireConfiguration().isEnabled()) {
+            toBeClosed.add(Services.createPGWireServer(
+                    config.getPGWireConfiguration(),
+                    workerPoolManager,
+                    cairoEngine,
+                    ffCache,
+                    snapshotAgent,
+                    metrics
+            ));
+        }
+
+        // ilp/tcp
+        toBeClosed.add(Services.createLineTcpReceiver(
+                config.getLineTcpReceiverConfiguration(),
+                workerPoolManager,
+                cairoEngine,
+                metrics
+        ));
+
+        // ilp/udp
+        toBeClosed.add(Services.createLineUdpReceiver(
+                config.getLineUdpReceiverConfiguration(),
+                sharedPool,
+                cairoEngine
+        ));
+
+        // telemetry
+        if (!cairoConfig.getTelemetryConfiguration().getDisableCompletely()) {
+            final TelemetryJob telemetryJob = new TelemetryJob(cairoEngine, ffCache);
+            toBeClosed.add(telemetryJob);
+            if (cairoConfig.getTelemetryConfiguration().getEnabled()) {
+                sharedPool.assign(telemetryJob);
+            }
+        }
+
+        System.gc(); // GC 1
+        log.advisoryW().$("Bootstrap complete, ready to start").$();
+    }
+
+    public void start() {
+        start(false);
+    }
+
+    public void start(boolean addShutdownHook) {
+        if (hasStarted.compareAndSet(false, true)) {
+            if (addShutdownHook) {
+                addShutdownHook();
+            }
+            workerPoolManager.startAll(log); // starts QuestDB's workers
+            Bootstrap.logWebConsoleUrls(config, log);
+            System.gc(); // final GC
+            log.advisoryW().$("enjoy").$();
+        }
+    }
+
+    @Override
+    public void close() {
+        if (hasStarted.compareAndSet(true, false)) {
+            workerPoolManager.closeAll();
+            ShutdownFlag.INSTANCE.shutdown();
+            Misc.freeObjList(toBeClosed);
+            toBeClosed.clear();
+        }
+    }
+
+    private void addShutdownHook() {
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            try {
+                System.err.println("QuestDB is shutting down...");
+                close();
+            } catch (Error ignore) {
+                // ignore
+            } finally {
+                LogFactory.INSTANCE.flushJobsAndClose();
+                System.err.println("QuestDB is shutdown.");
+            }
+        }));
+    }
+
+    public static void main(String[] args) throws Exception {
+        try {
+            new Server(args).start(true);
+        } catch (Throwable thr) {
+            System.err.println(thr.getMessage());
+            LogFactory.INSTANCE.flushJobsAndClose();
```

Comment:
this is an odd requirement to be aware of the logger. `Server` does not give caller opportunity to init logger. Bootstrap does that explicitly. It should also tidy up should anything go wrong